### PR TITLE
(WIP) Terraform 0.12x Support (Take Two)

### PR DESCRIPTION
**Pull request by jeffmccune**
This is a continuation of the work started in #198 

Tracked in #172

### CODE REVIEW

**Comment by jeffmccune** [comment]
Gary and I left off having fixed the networks input variables in https://github.com/forseti-security/terraform-google-forseti/pull/202/commits/fe7368fb174cfebfa6ef676d7619fb961a6b8967

We left the day trying to troubleshoot this error:

```
Error: Error patching router us-central1/forseti-simple-example-midge: googleapi: Error 400: Invalid value for field 'resource.nats[0].subnetworks[0].sourceIpRangesToNat': ''. The subnetwork is already configured with ALL_IP_RANGES for Nat which conflicts with the specified IpRangeToNatOption options in this Nat., invalid

  on ../../../examples/real_time_enforcer/nat.tf line 41, in resource "google_compute_router_nat" "main":
  41: resource "google_compute_router_nat" "main" {
```

Here's the build log location.  We're sort of stumped, this may be an issue with the provider?

https://concourse.infra.cft.tips/teams/cft/pipelines/terraform-google-forseti/jobs/integration-tests/builds/519#L5d1c3fba:365

File: modules/client/variables.tf [hunk 1]
```diff
@@ -79,26 +79,27 @@ variable "server_address" {
...
+  type        = list(string)
   default     = []
 }
 
+
 variable "client_access_config" {
   description = "Client instance 'access_config' block"
-  type        = "map"
-  default     = {}
+  default     = null
```
Comment (aaron-lane) [inline modules/client/variables.tf]: The empty map deafult should be preserved.

File: modules/real_time_enforcer/outputs.tf [hunk 2]
```diff
@@ -16,20 +16,26 @@
...
+  value       = google_compute_instance.main.name
 }
 
 output "forseti-rt-enforcer-vm-ip" {
   description = "Forseti Enforcer VM private IP address"
-  value       = "${google_compute_instance.main.network_interface.0.network_ip}"
+  value       = google_compute_instance.main.network_interface[0].network_ip
+}
+
+output "forseti-rt-enforcer-vm-public-ip" {
```
Comment (aaron-lane) [inline modules/real_time_enforcer/outputs.tf]: This change reverts da1a4aa8f01c91df36c3fbd1e2aee7c77d8abdea; let's drop it.

File: modules/real_time_enforcer/variables.tf [hunk 3]
```diff
@@ -70,20 +70,24 @@ variable "network_project" {
...
-  type        = "map"
+  type        = map(string)
   default     = {}
 }
 
 variable "enforcer_instance_access_config" {
   description = "Enforcer instance 'access_config' block"
-  type        = "map"
-  default     = {}
+  default     = null
```
Comment (aaron-lane) [inline modules/real_time_enforcer/variables.tf]: The empty map default should be preserved.

File: modules/real_time_enforcer/variables.tf [hunk 4]
```diff
@@ -70,20 +70,24 @@ variable "network_project" {
...
+  type = object({
```
Comment (aaron-lane) [inline modules/real_time_enforcer/variables.tf]: Should this also be `map(any)` to correspond with the similar change to the other access configuration variables?

File: modules/server/variables.tf [hunk 5]
```diff
@@ -544,20 +557,20 @@ variable "server_boot_image" {
...
-  type        = "list"
+  type        = list(string)
   default     = []
 }
 
 variable "server_access_config" {
   description = "Server instance 'access_config' block"
-  type        = "map"
-  default     = {}
+  default     = null
```
Comment (aaron-lane) [inline modules/server/variables.tf]: The default of an empty map should be preserved.

File: variables.tf [hunk 6]
```diff
@@ -76,32 +76,32 @@ variable "server_boot_image" {
...
-  type        = "list"
+  type        = list(string)
   default     = []
 }
 
 variable "server_access_config" {
   description = "Server instance 'access_config' block"
-  type        = "map"
-  default     = {}
+  default     = null
```
Comment (aaron-lane) [inline variables.tf]: The empty map default should be preserved.

File: variables.tf [hunk 7]
```diff
@@ -627,26 +642,26 @@ variable "client_region" {
...
-  type        = "list"
+  type        = list(string)
   default     = []
 }
 
 variable "client_access_config" {
   description = "Client instance 'access_config' block"
-  type        = "map"
-  default     = {}
+  default     = null
```
Comment (aaron-lane) [inline variables.tf]: The empty map default should be preserved.

**Comment by glarizza** [comment]
The `shared-vpc-local` test suite verified successfully on my local machine: https://gist.github.com/glarizza/8360245985066e342cfbc00b3cd2554c

**Comment by glarizza** [comment]
And `simple-example-local` is also verifying successfully on my local machine: https://gist.github.com/glarizza/d803cacc2eec5077bd501953e55c9cce

File: .envrc [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# export CREDENTIALS_PATH="${GOOGLE_CREDENTIALS_PATH}"
```
Comment (aaron-lane) [inline .envrc]: This file should be excluded.

**Pull request merged**